expr: `pnpm` monorepo

peetzweg opened the pull request:
Exchanges `deno` with traditional `node` tooling.

[`pnpm`](https://pnpm.io/) is a widely used supported package manager for node, which offers lean monorepo capabilities.

It ships with `node>=16` and can be enable with `corepack` which is part of the node installation.
```
corepack enable
```

The `pnpm-workspace.yaml` defines the path to individual projects of the monorepo.

https://github.com/paritytech/useink/blob/5df27adcf0cc110965a097a77ca9acc04d50e8c3/pnpm-workspace.yaml#L1-L3

Each project can have it's own node setup.

Use `pnpm install` at the root to install all dependencies of defined monorepo projects.
```
pnpm install
```

I added some convenience scripts in the root package.json to build a the packages or to start a development environment running the "kitchen-sink" frontend.

https://github.com/paritytech/useink/blob/ac11bff016b4484f8aad71d5294a9e8dd16a8b89/package.json#L6-L9

Build packages:
```
pnpm build
```

Run development setup:
```
pnpm dev
```

The development setup runs the `tsup` tooling in watch mode and starts the kitchen sink nextjs setup in development mode for recompilation on file changes.

---

[`tsup`](https://github.com/egoist/tsup) is a popular choice for bundling ts libraries.
[tsup docs](https://tsup.egoist.dev/)

---

The setup was heavily inspired by the repo setup of [wagmi](https://github.com/wagmi-dev/wagmi). However, I have great experience using these tools in the past.


### Why?

- I struggled with compiling useink with deno in "watch" mode for speeding up development cycles
- [`dnt` (Deno Node Transform)](https://github.com/denoland/dnt) was super slow for me and I don't want to invoke it manually all the time.
- `dnt` states itself it's under early development and may not be suitable for production: `This tool is under active early development and hasn't been tested in a lot of scenarios.`

- we are catering to the node world first and foremost. We are building a frontend library here, `deno` nor `bun` are not as widely used for this just yet. IMHO I don't find it worthwhile to drop all the existing mature tooling and struggling with comparability for {insert deno advantage here}. 🤷

peetzweg commented:
Hey @DoubleOTheven, can you give this a try and see if this works for you.
Just updated this to your recent changes in kitchen sink and useink as well.


As mentioned in the PR description I think it has some key advantages over the current deno setup. Please let me know what you think, if we go with this monorepo node approach I will adapt the gh actions as well.

DoubleOTheven commented:
This is a great change. I love that we get hot reloading and have a mono repo. 

As per our call, next steps are to wrap up the PRs for Chaindata and Notifications, then move to this approach.

Nice work!

DoubleOTheven reviewed (changes requested):
Nice changes. Let's get the new Chain config and Notification PRs in first, then update this PR.

nicolad commented:
Amazing direction!
At current FT job I've implemented `pnpm` along `turborepo` to simplify overall build process:

https://turbo.build/repo

peetzweg commented:
> Amazing direction! At current FT job I've implemented `pnpm` along `turborepo` to simplify overall build process:
> 
> https://turbo.build/repo

Yes that would be the next step. Not sure if turborepo or nx. I personally I'm a fan and user of nx but haven't tried turborepo in the last 8 months or so. We got it on our radar for further improvement of the development experience.

nicolad commented:
heya 👋 
what the progress on this, I need this to ease my testing regarding this issue:
https://github.com/paritytech/useink/issues/40

if there are any ways I can facilitate this, let me know

DoubleOTheven commented:
Version 1.1.0 has been released. Do you want to look into migrating to node based off of the latest main branch?

peetzweg commented:
Yep will give it another go. 👍:)

peetzweg commented:
Just updated the code in this branch to the most recent `useink` version.

Fyi it does not have multiple packages as the original deno code doesn't. It's just multiple entry points in a single packages as the deno version did it.

So for this initial switch we only have two projects in this this monorepo workspace, `packages/useink` as it's working before and the `useink-kitchen-sink` as `playground`.

Please give it a try @DoubleOTheven if everything works as expected. Also happy to get some feedback from @nicolad and @felixpolimec (#46).

At best fresh copy of this branch then use `pnpm i` to install all dependencies and run the playground and bundling of the library with `pnpm dev`.

nicolad commented on playground/.nvmrc:
```diff
@@ -0,0 +1 @@
+19
```
is `.nvmrc` needed only for playground or the whole monorepo setup?

nicolad commented:
hey @peetzweg thanks for giving it another go 🚀 
I am using node `v18.12.0` currently and it works perfectly, aside those conflicting errors you pointed out.
Switched to v19 using nvm, works perfectly as well.
I had some similar errors while using turborepo + pnpm and fixed with [`pnpm.overrides`](https://pnpm.io/package_json#pnpmoverrides), not sure if the same fix might apply here, but worth giving a try.

DoubleOTheven commented:
Thanks this change @peetzweg. It works great on my end. The watch mode is 🔥

DoubleOTheven reviewed (changes requested):
I would love to merge this in. Just two requests.

1. Please pull in the latest changes from `1.4.2` and `1.4.3`
2. We lost linting and auto formatting. Deno has this by default. Would you please add prettier and eslint? I'm open to moving that task to another PR if you think it best. Just raising the need...

peetzweg commented:
> I would love to merge this in. Just two requests.
> 
>     1. Please pull in the latest changes from `1.4.2` and `1.4.3`
> 
>     2. We lost linting and auto formatting. Deno has this by default. Would you please add prettier and eslint? I'm open to moving that task to another PR if you think it best. Just raising the need...

Great points! Just updated the code. 

Instead of prettier and eslint I've removed them from the `playground` project and added [`rome`](https://rome.tools/) instead. It got quite popular and I think it's worth a try. Works great so far. Not so many languages or plugins supported but we got what we need so far.

https://github.com/paritytech/useink/blob/72fbb5c12d027b736bf3653ba384893efb725377/package.json#L11-L13

Furthermore, I've added a `settings.json` and `extensions.json` file for the for `vscode`. Suggestion some extensions and enabling some workspace setting to get up and running with rome.

https://github.com/paritytech/useink/blob/72fbb5c12d027b736bf3653ba384893efb725377/.vscode/settings.json#L1-L7

https://github.com/paritytech/useink/blob/72fbb5c12d027b736bf3653ba384893efb725377/.vscode/extensions.json#L1-L4

DoubleOTheven commented on packages/useink/useink-1.4.4.tgz:
Did you mean to commit this?

DoubleOTheven reviewed (changes requested):
These changes all look good, but there are a lot of collisions with the main branch. Would you please resolve them and make this PR a single commit with `chore: change Deno to node and pnpm` (or something similar?)

I tried doing rebasing but there are too many changes. You might be better off converting mod.ts to index.ts as you did the first time.

peetzweg commented:
@DoubleOTheven 

merged with main to resolve conflicts. Fixed lint and spellcheck actions. Had to fix some lint issues so lint ci will succeed. This commit includes the fixes:
https://github.com/paritytech/useink/pull/55/commits/071d27cac79229cb20acdf463e66c421d4dee7cb

Release action should work as well 🤞

DoubleOTheven commented on packages/useink/src/react/hooks/contracts/useEventSubscription.ts:
```diff
@@ -0,0 +1,74 @@
+import { Bytes } from '../../../core/index';
```
This file name should be `useEvents`. It was changed
